File: Permutations/I/Permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = [[]]
        self._DFS(nums, 0, result)
        return result

    def _DFS(nums, begin, result):
    	if now == len(nums):
    		result.append(list(nums))
    		return
    	for i in xrange(begin, len(nums)):
    		nums[i], nums[begin] = nums[begin], nums[i]
    		self._DFS(nums, begin + 1, result)
    		nums[i], nums[begin] = nums[begin], nums[i]
    	return

File: Permutations/I/Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        DFS(nums, 0, result);
        return result;
    }
private:
	void DFS(vector<int> &nums, int begin, vector<vector<int>> &result) {
		if(begin == nums.size()) {
			result.push_back(nums);
			return;
		}
		for(int i = begin; i < nums.size(); i++) {
			swap(nums[i], nums[begin]);
			DFS(nums, begin + 1, result);
			swap(nums[i], nums[begin]);
		}
	}
};